From 7810361dbbcffee39fed33a9f4cff7eff3b3ea1f Mon Sep 17 00:00:00 2001
From: yangjunyi <yangjunyi5@huawei.com>
Date: Mon, 21 Nov 2022 10:39:05 +0800
Subject: [PATCH 0081/1398] [Huawei] drivers: hooks: add libata driver vendor
 hooks

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187991

----------------------------------

Allow vendor modules to attach libata driver hooks. This
patch introduce vendor_sata_reset hook.

Usage:

  static void vendor_libata_foo(void *data, struct ata_link *link)
  {
          pr_info("%s\n", __func__);
  }

  static int __init vendor_init(void)
  {
	int ret = 0;
	ret += register_trace_vendor_sata_reset(&vendor_libata_foo, NULL);

	return ret;
  }

  static void __exit vendor_exit(void)
  {
  	unregister_trace_vendor_sata_reset(&vendor_libata_foo, NULL);
  }

  module_init(vendor_init);
  module_exit(vendor_exit);

Signed-off-by: yangjunyi <yangjunyi5@huawei.com>
---
 drivers/ata/libata-eh.c      |  2 ++
 drivers/hooks/Kconfig        | 10 ++++++++++
 drivers/hooks/vendor_hooks.c |  5 +++++
 include/trace/hooks/libata.h | 35 +++++++++++++++++++++++++++++++++++
 4 files changed, 52 insertions(+)
 create mode 100644 include/trace/hooks/libata.h

diff --git a/drivers/ata/libata-eh.c b/drivers/ata/libata-eh.c
index 0cd5e36d36be..07dfb71db69c 100644
--- a/drivers/ata/libata-eh.c
+++ b/drivers/ata/libata-eh.c
@@ -26,6 +26,7 @@
 #include <linux/libata.h>
 
 #include <trace/events/libata.h>
+#include <trace/hooks/libata.h>
 #include "libata.h"
 
 enum {
@@ -2430,6 +2431,7 @@ static int ata_do_reset(struct ata_link *link, ata_reset_fn_t reset,
 {
 	struct ata_device *dev;
 
+	trace_vendor_sata_reset(link);
 	if (clear_classes)
 		ata_for_each_dev(dev, link, ALL)
 			classes[dev->devno] = ATA_DEV_UNKNOWN;
diff --git a/drivers/hooks/Kconfig b/drivers/hooks/Kconfig
index 6a00168e67ad..9eaebaa17a97 100644
--- a/drivers/hooks/Kconfig
+++ b/drivers/hooks/Kconfig
@@ -20,4 +20,14 @@ config VENDOR_BOND_HOOKS
           Allow vendor modules to attach bonding driver hooks defined via
           DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
 
+config VENDOR_LIBATA_HOOKS
+        bool "Libata driver Vendor Hooks"
+        depends on VENDOR_HOOKS && ATA
+        default n
+        help
+        Enable libata driver vendor hooks
+
+        Allow vendor modules to attach libata driver hooks defined via
+        DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
+
 endmenu
diff --git a/drivers/hooks/vendor_hooks.c b/drivers/hooks/vendor_hooks.c
index 85bda58159f6..102ca5397935 100644
--- a/drivers/hooks/vendor_hooks.c
+++ b/drivers/hooks/vendor_hooks.c
@@ -9,6 +9,7 @@
 #define CREATE_TRACE_POINTS
 #include <trace/hooks/vendor_hooks.h>
 #include <trace/hooks/bonding.h>
+#include <trace/hooks/libata.h>
 
 /*
  * Export tracepoints that act as a bare tracehook (ie: have no trace event
@@ -18,3 +19,7 @@
 #ifdef CONFIG_VENDOR_BOND_HOOKS
 EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_bond_check_dev_link);
 #endif
+#ifdef CONFIG_VENDOR_LIBATA_HOOKS
+/* power on/off sata */
+EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_sata_reset);
+#endif
diff --git a/include/trace/hooks/libata.h b/include/trace/hooks/libata.h
new file mode 100644
index 000000000000..6c8469f61b77
--- /dev/null
+++ b/include/trace/hooks/libata.h
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
+ * Description: Libata driver Vendor Hooks
+ */
+
+#ifdef CONFIG_VENDOR_LIBATA_HOOKS
+#undef TRACE_SYSTEM
+#define TRACE_SYSTEM libata
+
+#define TRACE_INCLUDE_PATH trace/hooks
+#if !defined(_TRACE_HOOK_LIBATA_HOOKS_H) || defined(TRACE_HEADER_MULTI_READ)
+#define _TRACE_HOOK_LIBATA_HOOKS_H
+#include <linux/tracepoint.h>
+#include <trace/hooks/vendor_hooks.h>
+
+struct ata_link;
+
+DECLARE_HOOK(vendor_sata_reset,
+	TP_PROTO(struct ata_link *link),
+	TP_ARGS(link));
+
+#endif
+
+/* This part must be outside protection */
+#include <trace/define_trace.h>
+
+#else
+struct ata_link;
+
+DECLARE_EVENT_NOP(vendor_sata_reset,
+	TP_PROTO(struct ata_link *link),
+	TP_ARGS(link));
+
+#endif
-- 
2.17.1

